Fix error resposes for `from` search prefix

aescling opened the pull request:
Addresses #17941.

Using unsupported prefixes now reports a 422; searching for posts from an account the instance is not aware of reports a 404. I think the UI for this on the front end is only barely more helpful at actually communicating any problems to the user, but at least the semantics are reasonable.

In addition, this commit addresses two minor points. More importantly, `from:username@domain` succeeds when `domain` is the local domain. Less importantly, `from:@username(@domain)?` is now valid syntax.

For the record, a nealy identical version of this commit in https://github.com/kibicat/mastodon/commit/945c9b1342b7a000da05074e4572abe6ce7feb9c is running on https://glitch.cat.family, with no reported issues.

aescling committed:
Fix error responses in `from` search prefix (addresses mastodon/mastodon#17941)

Using unsupported prefixes now reports a 422; searching for posts from an
account the instance is not aware of reports a 404. TODO: The UI for this
on the front end is abysmal.

Searching `from:username@domain` now succeeds when `domain` is the local
domain; searching `from:@username(@domain)?` now works as expected.

* Remove unused methods on new Error classes as they are not being used

Currently when `raise`d there are error messages being supplied, but
this is not actually being used. The associated `raise`s have been
edited accordingly.

* Remove needless comments

* Satisfy rubocop

* Try fixing tests being unable to find AccountFindingConcern methods

* Satisfy rubocop

ClearlyClaire reviewed (changes requested):
Thanks for the contribution!

The change is welcome even though I would have done without additional helpers and duplicate exceptions (`Mastodon::NotFound` could just be `ActiveRecord::RecordNotFound`).

single-right-quote commented on app/lib/search_query_transformer.rb:
```diff
@@ -88,14 +88,16 @@ def initialize(prefix, term)
-        username, domain = term.split('@')
-        account = Account.find_remote(username, domain)
 
-        raise "Account not found: #{term}" unless account
+        username, domain = Account.validate_account_string(term)
+        raise Mastodon::SyntaxError if username.nil?
+
+        account = Account.find_local_or_remote(username, domain)
+        raise Mastodon::NotFound if account.nil?
```
I&#x2019;m not familiar with project etiquette here. I have recreated your suggestion locally (and removed code I added that it renders unused), but should I instead commit your suggestion from GitHub, then removed the needless extra code. and then squash?

ClearlyClaire replied:
Honestly, do as is more convenient for you (maybe avoiding force-pushes though). The PRs get squashed once they get accepted, so it is not necessary for you to do so.

single-right-quote replied:
Thank you; pushed.

ClearlyClaire commented:
Honestly, do as is more convenient for you (maybe avoiding force-pushes though). The PRs get squashed once they get accepted, so it is not necessary for you to do so.

aescling committed:
Simplify `from` prefix logic

This incorporates @ClearlyClaire's suggestion (see
https://github.com/mastodon/mastodon/pull/17963#pullrequestreview-933986737).

Accepctable account strings in `from:` clauses are more lenient than
before this commit; for example, `from:@user@example.org@asnteo +cat`
will not error, and return posts by @user@example.org containing the
word "cat". This is more consistent with how Mastodon matches mentions
in statuses. In addition, `from` clauses will not be checked for
syntatically invalid usernames or domain names, simply 404ing when
`Account.find_remote!` raises ActiveRecord::NotFound.

New code for this PR that is no longer used has been removed.

ClearlyClaire reviewed (approved):
I'd rather have an “invalid” prefix either silently fall back to textual comparison, or have a more informative error, but it's already a noticeable improvement over the current situation.

single-right-quote commented:
I agree. An example where this can be a problem is searching for links, since `https://` is interpreted as a prefix, so either special casing that to fall back (ew) or just falling back in general seems very much worth exploring to me.
